README.md
# RSS Spider

[中文](https://github.com/Neutrino3316/rss_spider/blob/master/README-zh.md) | English

[CHANGELOG](https://github.com/Neutrino3316/rss_spider/blob/master/CHANGELOG.md)

RSS Spider is a repo aim to fetch and save data from known RSS source.(is NOT fetch data from normal website and generate RSS source)

The reasons of this repo are:

1. RSS provides formatted data, very easy for us to fetch and save data
2. The RSS source generate new data continuously, I hope I can deposit these data in my own database.
3. If I crawl every website on my own, I have to write different spider for different website. A way to reduce my workload, I can use the RSS generated by [RSSHub](https://github.com/DIYgod/RSSHub), which has many codes to fetch data from different website.
4. By setting up a docker container, I can easily grab data all the time, and it's convenient to manage it.

This repo is still working in progress, any suggestions, issues and pull requests are warmly welcomed.

# Project Structure

## File Structure

- config.yml: this file is used to pass the settings
- run.py: to run the whole project

# Configuration

It is highly recommended to use docker to run this project.

## config.yml

An example config.yml is as the following:

```yaml
mongodb:
    link: mongodb://localhost:27017
rsshub:
    host: http://localhost:1200/
rss:
    zhihu_hotlist:
        link: https://rsshub.app/zhihu/hotlist
        key_list:
            - title
            - link
            - published
            - author
            - summary
```

- Line 2: the address of mongoDB database. If you use docker link container, please use the mongo container name (e.g. mongo_rss_spider) to replace localhost
- Line 4: the address of RSSHub, please specify http or https, this address will auto replace `rsshub_host` in the links of rss. If you have your own RSSHub, please use the IP address and the port of the RSSHub container, if you don't have your own RSSHub, please use `https://rsshub.app/`. If you use docker link container, please use the RSSHub container name (e.g. rsshub_diygod) to replace localhost
- Line 6: zhihu_hotlist is the name of a sub-project. It will generate a collection named zhihu_hotlist in rss_spider database.
- Line 7: the link to fetch RSS
- Line 9 to line 13: the keys to store RSS, this key should match the keys in this RSS. Title and link are necessary keys.

For the time being, only mongoDB database is supported.

## run your own RSSHub (optional)

It's highly recommended to run your own RSSHub. On one hand, it can speed up the refresh rate, on the other hand, it can reduce the work load of the sever own by the RSSHub author.

You may deploy your own rsshub using [the offical online document of RSSHub](https://docs.rsshub.app/en/install/).

I use the following command to run rsshub in docker, I reduce the cache time, so that it can refresh more quickly.

```bash
docker run -d --name rsshub_diygod --restart=always -p 1200:1200 \
	-e CACHE_EXPIRE=5 -e CACHE_CONTENT_EXPIRE=60 \
	diygod/rsshub:latest
```

## run your own mongoDB in docker (optional)

```bash
docker run -d -p 27017:27017 --name mongo_rss_spider --restart=always \
	-v mongo_rss_spider_data_configdb:/data/configdb \
	-v mongo_rss_spider_data_db:/data/db \
	-v d:/docker_mount/mongo_rss_spider_backup:/mongo_backup \
	mongo
```

# Acknowledgement

Many thanks to the project [RSSHub](https://github.com/DIYgod/RSSHub) which is written by [DIYgod](https://github.com/DIYgod)



CHANGELOG.md
# 1.0.0 (future)

- **Initial Release**

# 0.2.1 (2020-02-05)

## Enhancement

- **Docker auto Publish Update**: Update the Github Action, now only pushes on master and development branch will trigger auto build and publish in DockerHub.

## Documentation

- [#23](https://github.com/neutrino3316/rss_spider/issues/23)  **CHANGELOG**: Add changelog in both English and Chinese.

# 0.2.0 (2020-02-05)

## Enhancement

- [#4](https://github.com/neutrino3316/rss_spider/issues/4)  **Logging**: Use logging instead of printing.
- [#29](https://github.com/neutrino3316/rss_spider/issues/29)  **Test Argument**: Add test argument in run.py

## Documentation

- [#25](https://github.com/neutrino3316/rss_spider/issues/25)  **Comments**: Add necessary comments.
- [#26](https://github.com/neutrino3316/rss_spider/issues/26)  **Project Structure**: Add illustration of project structure.
- [#28](https://github.com/neutrino3316/rss_spider/issues/28)  **Push Readme to DockerHub**: Add github actions that can push readme to DockerHub.

# 0.1.6 and before (2019-12-18)

- **Basic code**: Building the foundation of this project.


test.py
"""
To find a better way to use multiprocessing, and logging within different process
"""

import multiprocessing
import time
import random
import logging
from logging.handlers import QueueHandler, QueueListener


class Worker(multiprocessing.Process):

    def __init__(self, worker_id: int, queue):
        multiprocessing.Process.__init__(self)
        self.worker_id = worker_id
        self.queue = queue

    def _setup_logger(self):

        self.queue_handler = QueueHandler(self.queue)
        self.logger = logging.getLogger("main.worker_%d" % self.worker_id)
        # self.logger.addHandler(self.queue_handler)
        self.logger.info("logger setup worker %d" % self.worker_id)

    def run(self):
        time.sleep(random.uniform(0.1, 0.5))
        self._setup_logger()
        # print("show_id called, worker_%d" % self.worker_id)
        # print(self.logger.handlers)
        self.logger.info("message from worker %d" % self.worker_id)
        time.sleep(random.uniform(0.1, 0.5))


def logger_init():
    queue = multiprocessing.Queue()
    # this is the handler for all log records
    handler = logging.StreamHandler()
    handler.setFormatter(logging.Formatter(
        "%(asctime)s %(levelname)s %(filename)s:%(lineno)d PID:%(process)d %(name)s \t %(message)s"
    ))

    # queue_listener gets records from the queue and sends them to the handler
    queue_listener = QueueListener(queue, handler)
    queue_listener.start()

    logger = logging.getLogger("main")
    logger.setLevel(logging.DEBUG)
    # add the handler to the logger so records from this process are handled
    logger.addHandler(handler)

    return queue_listener, queue


if __name__ == '__main__':

    queue_listener, queue = logger_init()

    logger = logging.getLogger("main.root")
    # logger.addHandler(QueueHandler(queue))
    print(logger.parent.handlers)
    print(logger.handlers)
    logger.info("logging from main, starting.")

    worker_num = 7
    worker_list = []
    for i in range(worker_num):
        worker_list.append(Worker(i, queue))

    for worker in worker_list:
        worker.run()

    for worker in worker_list:
        worker.join()

    queue_listener.stop()
    logger.info("logging from main, everything ended.")


test2.py
"""sample code for logging in subprocesses using multiprocessing

* Little handler magic - The main process uses loggers and handlers as normal.
* Only a simple handler is needed in the subprocess that feeds the queue.
* Original logger name from subprocess is preserved when logged in main
  process.
* As in the other implementations, a thread reads the queue and calls the
  handlers. Except in this implementation, the thread is defined outside of a
  handler, which makes the logger definitions simpler.
* Works with multiple handlers.  If the logger in the main process defines
  multiple handlers, they will all be fed records generated by the
  subprocesses loggers.

tested with Python 2.5 and 2.6 on Linux and Windows

"""

import os
import sys
import time
import traceback
import multiprocessing, threading, logging, sys
from logging.handlers import QueueHandler

DEFAULT_LEVEL = logging.DEBUG

formatter = logging.Formatter("%(levelname)s: %(asctime)s - %(name)s - %(process)s - %(message)s")

class LogQueueReader(threading.Thread):
    """thread to write subprocesses log records to main process log

    This thread reads the records written by subprocesses and writes them to
    the handlers defined in the main process's handlers.

    """

    def __init__(self, queue):
        threading.Thread.__init__(self)
        self.queue = queue
        self.daemon = True

    def run(self):
        """read from the queue and write to the log handlers

        The logging documentation says logging is thread safe, so there
        shouldn't be contention between normal logging (from the main
        process) and this thread.

        Note that we're using the name of the original logger.

        """
        # Thanks Mike for the error checking code.
        while True:
            try:
                record = self.queue.get()
                # get the logger for this record
                logger = logging.getLogger(record.name)
                logger.callHandlers(record)
            except (KeyboardInterrupt, SystemExit):
                raise
            except EOFError:
                break
            except:
                traceback.print_exc(file=sys.stderr)

class LoggingProcess(multiprocessing.Process):

    def __init__(self, queue):
        multiprocessing.Process.__init__(self)
        self.queue = queue

    def _setupLogger(self):
        # create the logger to use.
        logger = logging.getLogger('test.subprocess')
        # The only handler desired is the SubProcessLogHandler.  If any others
        # exist, remove them. In this case, on Unix and Linux the StreamHandler
        # will be inherited.

        for handler in logger.handlers:
            # just a check for my sanity
            assert not isinstance(handler, QueueHandler)
            logger.removeHandler(handler)
        # add the handler
        handler = QueueHandler(self.queue)
        handler.setFormatter(formatter)
        logger.addHandler(handler)

        # On Windows, the level will not be inherited.  Also, we could just
        # set the level to log everything here and filter it in the main
        # process handlers.  For now, just set it from the global default.
        logger.setLevel(DEFAULT_LEVEL)
        self.logger = logger

    def run(self):
        self._setupLogger()
        logger = self.logger
        # and here goes the logging
        p = multiprocessing.current_process()
        logger.info('hello from process %s with pid %s' % (p.name, p.pid))


if __name__ == '__main__':
    # queue used by the subprocess loggers
    queue = multiprocessing.Queue()
    # Just a normal logger
    logger = logging.getLogger('test')
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(DEFAULT_LEVEL)
    logger.info('hello from the main process')
    # This thread will read from the subprocesses and write to the main log's
    # handlers.
    log_queue_reader = LogQueueReader(queue)
    log_queue_reader.start()
    # create the processes.
    for i in range(10):
        p = LoggingProcess(queue)
        p.start()
    # The way I read the multiprocessing warning about Queue, joining a
    # process before it has finished feeding the Queue can cause a deadlock.
    # Also, Queue.empty() is not realiable, so just make sure all processes
    # are finished.
    # active_children joins subprocesses when they're finished.
    while multiprocessing.active_children():
        time.sleep(.1)

run.py
import argparse
import datetime
import dateutil.parser
import feedparser
import logging
import time
import yaml
from multiprocessing import Pool
from pymongo import MongoClient

# set logging format and logging level
# logging.basicConfig(format="%(asctime)s %(message)s", level=logging.DEBUG)
logging.basicConfig(format="%(asctime)s %(levelname)s %(filename)s:%(lineno)d PID:%(process)d %(name)s \t %(message)s",
                    level=logging.DEBUG)

# set time format
strftime_time_formatted = {
    "with_time_zone": "%y-%m-%d %H:%M:%S %Z",
    "without_time_zone": "%y-%m-%d %H:%M:%S %Z"}


def get_time_now():
    local_time = datetime.datetime.utcnow() + datetime.timedelta(hours=8)
    local_time = local_time.replace(tzinfo=datetime.timezone(datetime.timedelta(hours=8)))
    return local_time.strftime(strftime_time_formatted["with_time_zone"])


class RSS:
    """
    A listener of one rss source
    """
    def __init__(self, rss_name: str, rss_link: str, rss_keys_list: list, database_link, database_name: str):
        # the rss information, including the name, the link address, the key list.
        self.rss_name = rss_name
        self.rss_link = rss_link
        self.rss_keys_list = rss_keys_list

        # the database information, the data fetch from the rss will be stored in this database
        self.database_link = database_link
        self.database_name = database_name

        self.log_tag = "rss %s\t" % self.rss_name
        self.rss = None  # the rss object returned by the feedparser
        self.new_items_count = 0  # how many new items found in the latest fetch compared to the penultimate one.
        self.save_item_list = []  # the key of the items need to be fetched and saved
        self.waiting_time = 1  # initial waiting time between each data fetch, the time unit is second

    #     self._create_logger()
    #     # self.logger.warning("This rss has created.")
    #
    # def _create_logger(self):
    #     self.logger = logging.getLogger("rss:%s\t" % self.rss_name)
    #     self.logger.setLevel(logging.DEBUG)
    #     self.logger.info("Logger created.")

    def fetch(self):
        """
        Fetch the rss. Using the self.rss_link. Data fetched will be stored in a rss object, in self.rss
        :return: None
        """
        self.rss = feedparser.parse(self.rss_link)

    def parse_item(self):
        """
        Using the data fetched in self.rss, parse the data and store them in self.save_item_list
        :return:
        """
        self.save_item_list = []
        for raw_item in self.rss.entries:
            save_item = dict((key, raw_item[key]) for key in self.rss_keys_list)
            save_item["_id"] = raw_item["link"]
            save_item["last_update_time"] = get_time_now()
            # save_item["published_parsed"] = RSS.parse_time(save_item.pop("published"))
            save_item["published_parsed"] = RSS.parse_time(save_item["published"])
            save_item = dict(sorted(save_item.items()))
            self.save_item_list.append(save_item)

    @staticmethod
    def parse_time(raw_time_string):
        """
        Parse the time string
        :param raw_time_string: a string of time
        :return: a class datetime.datetime object
        """
        return dateutil.parser.parse(raw_time_string)

    def save_item(self):
        """
        Save the new items into the database
        Store the amount of new items in self.new_items_count
        Start and close a connection to the database
        :return: None
        """
        self.new_items_count = 0
        db_client = MongoClient(self.database_link)
        database = db_client[self.database_name]
        for save_item in self.save_item_list:
            update_result = database[self.rss_name].update_one({'_id': save_item['_id']}, {"$set": save_item},
                                                               upsert=True)
            if update_result.matched_count == 0:
                self.new_items_count += 1
                logging.info(self.log_tag + "Add new item:\n%s\n%s" % (save_item["title"], save_item["link"]))
        db_client.close()

    def update_waiting_time(self):
        """
        Algorithm of updating the waiting time, i.e. self.waiting_time
        Double the waiting time if no new items are found in the latest fetch.
        Cut the waiting time in half if more than 5 new items are found in the latest fetch.
        :return: None
        """
        if self.new_items_count == 0:
            self.waiting_time *= 2
            # self.logger.info("Waiting time doubled, the new waiting time is %d seconds" % self.waiting_time)
            logging.info(self.log_tag + "Waiting time doubled, the new waiting time is %d seconds." % self.waiting_time)
        elif self.new_items_count > 5:
            self.waiting_time /= 2
            # self.logger.info("Waiting time cut in half, the new waiting time is %d seconds" % self.waiting_time)
            logging.info(self.log_tag + "Waiting time cut in half, the new waiting time is %d seconds." % self.waiting_time)

    def run(self):
        """
        Start a dead loop. It will fetch and save the data continuously.
        :return: None
        """
        while True:
            # self.logger.info("Start now.")
            logging.info(self.log_tag + "Start a new round.")
            self.fetch()
            self.parse_item()
            self.save_item()
            self.update_waiting_time()
            # self.logger.info("start end, will waiting %d seconds before next run" % self.waiting_time)
            logging.info(self.log_tag + "End this round, will wait %d seconds before next round." % self.waiting_time)
            time.sleep(self.waiting_time)


if __name__ == "__main__":

    # create root logger
    logger = logging.getLogger("main")
    logger.setLevel(logging.DEBUG)

    logger.warning("The initial of this program, everything will load and start in a minute.")

    # read config.yml
    with open("config.yml", 'r') as ymlfile:
        config = yaml.load(ymlfile, Loader=yaml.SafeLoader)
    logger.info("Config loaded: %s" % config)

    # parse the arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("--mongodb_host", default=None, type=str,
                        help="the link of mongoDB, e.g. localhost:27017， mongo_rss_spider:27017")
    parser.add_argument("--rsshub_host", default=None, type=str,
                        help="the host of RSSHub with http or https, e.g. \
                        https://rsshub.app http://localhost:1200 http://rsshub_diygod:1200")
    parser.add_argument("--test", action="store_true",
                        help="if test, write data to database named rss_spider_test.")
    args = parser.parse_args()

    # override the mongodb_host and the rsshub_host if the corresponding argument is passed.
    logger.info("Arguments loaded: %s" % args)
    if args.mongodb_host is not None:
        config['mongodb']['link'] = args.mongodb_host
        logger.info("Using mongodb_host from args, i.e. %s" % args.mongodb_host)
    if args.rsshub_host is not None:
        config["rsshub"]["host"] = args.rsshub_host
        logger.info("Using rsshub_host from args, i.e. %s" % args.rsshub_host)

    # whether to save the data in test database or not
    if args.test:
        database_name = "rss_spider_test"
    else:
        database_name = "rss_spider"

    # for every rss, replace host name and sort items
    for rss_name in config["rss"].keys():
        config["rss"][rss_name]["link"] = config["rss"][rss_name]["link"].replace(
            "rsshub_host", config["rsshub"]["host"])
    config = dict(sorted(config.items()))

    # create a rss_list that include all rss
    rss_list = []
    for key, value in config["rss"].items():
        rss = RSS(key, value["link"], value["key_list"], config['mongodb']['link'], database_name)
        rss_list.append(rss)

    # print the number of rss loaded and all rss details
    logger.info("%d rss are loaded. They're list as the followings:" % len(rss_list))
    for i, rss in enumerate(rss_list):
        logger.info("%d/%d rss name: %s \t link: %s" % (i+1, len(rss_list), rss.rss_name, rss.rss_link))

    # use multiprocessing, for each rss, create a process that continuously fetch, parse and store data
    logger.warning("Start all rss. Each rss has a unique thread to fetch parse and store the data.")
    pool = Pool(len(rss_list))
    for i in range(len(rss_list)):
        pool.apply_async(rss_list[i].run)
    pool.close()
    pool.join()

    # should never be run
    logger.error("All ended!")
